### PR TITLE
[SPARK-23202][SQL] Add new API in DataSourceWriter: onDataWriterCommit 

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

The current DataSourceWriter API makes it hard to implement `onTaskCommit(taskCommit: TaskCommitMessage)` in `FileCommitProtocol`.
In general, on receiving commit message, driver can start processing messages(e.g. persist messages into files) before all the messages are collected.

The proposal to add a new API:
`add(WriterCommitMessage message)`:  Handles a commit message on receiving from a successful data writer.

This should make the whole API of DataSourceWriter compatible with `FileCommitProtocol`, and more flexible.

There was another radical attempt in #20386.  This one should be more reasonable.

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
@cloud-fan @jose-torres @rdblue

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataSourceWriter.java [hunk 1]
```diff
@@ -62,6 +62,15 @@
    */
   DataWriterFactory<Row> createWriterFactory();
 
+  /**
+   * Handles a commit message on receiving from a successful data writer.
+   *
+   * If this method fails (by throwing an exception), this writing job is considered to to have been
+   * failed, and {@link #abort(WriterCommitMessage[])} would be called. The state of the destination
+   * is undefined and {@link #abort(WriterCommitMessage[])} may not be able to deal with it.
```
Comment (rdblue) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataSourceWriter.java]: What does it mean that "the state of the destination is undefined"? I think it is sufficient to say that `abort` will be called and the contract for aborting commits applies.

**Comment by rdblue** [comment]
+1

I'd rather not add features without a known use case, but this implementation looks good to me.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataSourceWriter.java (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataSourceWriter.java]: makes sense, let's remove the last sentence.

**Comment by cloud-fan** [comment]
adding a default method to a java interface is binary compatible, I'm merging this to master only, to follow @rxin 's suggestion, thanks!

**Pull request closed without merging**